nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFTP.java

```diff
@@ -0,0 +1,170 @@
+import org.mockftpserver.fake.UserAccount;
+import org.mockftpserver.fake.filesystem.DirectoryEntry;
+import org.mockftpserver.fake.filesystem.FileEntry;
+import org.mockftpserver.fake.filesystem.FileSystem;
+import org.mockftpserver.fake.filesystem.WindowsFakeFileSystem;
+
+import java.io.FileInputStream;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+public class TestFTP {
+
+    final FakeFtpServer fakeFtpServer = new FakeFtpServer();
+    final String username = "nifi-ftp-user";
+    final String password = "Test test test chocolate";
+    int ftpPort;
+
+    @Before
+    public void setUp() throws Exception {
+        fakeFtpServer.setServerControlPort(0);
+        fakeFtpServer.addUserAccount(new UserAccount(username, password, "c:\\data"));
+
+        FileSystem fileSystem = new WindowsFakeFileSystem();
+        fileSystem.add(new DirectoryEntry("c:\\data"));
+        fakeFtpServer.setFileSystem(fileSystem);
+
+        fakeFtpServer.start();
+
+        ftpPort = fakeFtpServer.getServerControlPort();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        fakeFtpServer.stop();
+    }
+
+    @Test
+    public void testValidators() {
+        TestRunner runner = TestRunners.newTestRunner(PutFTP.class);
+        Collection<ValidationResult> results;
+        ProcessContext pc;
+
+        /* Set the basic required values */
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.USERNAME, "${el-username}");
+        runner.setProperty(FTPTransfer.HOSTNAME, "static-hostname");
+        runner.setProperty(FTPTransfer.PORT, "${el-portNumber}");
+
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.REMOTE_PATH, "static-remote-target");
+        runner.enqueue(new byte[0]);
+        pc = runner.getProcessContext();
+        if (pc instanceof MockProcessContext) {
+            results = ((MockProcessContext) pc).validate();
+        }
+        assertEquals(0, results.size());
+
+
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.REMOTE_PATH, "${el-remote-target}");
+        runner.enqueue(new byte[0]);
+        pc = runner.getProcessContext();
+        if (pc instanceof MockProcessContext) {
+            results = ((MockProcessContext) pc).validate();
+        }
+        assertEquals(0, results.size());
+
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.USERNAME, "static-username");
+        runner.enqueue(new byte[0]);
+        pc = runner.getProcessContext();
+        if (pc instanceof MockProcessContext) {
+            results = ((MockProcessContext) pc).validate();
+        }
+        assertEquals(0, results.size());
+
+        /* Try an invalid expression */
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.USERNAME, "$static-username}");
```

Comment:
In the test, `MockValidationContext` is used which does not validate the actual EL expression. Therefore, I was getting a test failure here (**0** validation errors present when **1** was expected). I changed the property value to be `""` and the test succeeded (i.e. the validator threw an error).